### PR TITLE
pg sources: configurable TCP keepalives

### PR DESCRIPTION
**Pull request by sploiselle**
Based on https://github.com/MaterializeInc/materialize/issues/17598#issuecomment-1498500020, we want to add TCP keepalives to notice when the connection has failed more eagerly than we were before.

@pH14 could you PTAL?

@MaterializeInc/qa Is there an existing pattern for only running `mzcompose` tests only on nightly? Glad to make a pattern but assumed there was something that I just couldn't find, so didn't want to make something needlessly duplicative.

### Motivation

This PR adds a known-desirable feature. https://github.com/MaterializeInc/materialize/issues/17598#issuecomment-1498500020

Note that this PR also fixes a bug; `pg-cdc-resumption` was not previously using the toxiproxy connection whatsoever, so was not successfully testing that we were resilient to interruption. Ran locally and it worked still, but this should likely get fixed quickly––will break this off into its own PR if things drag.

### Tips for reviewer

Note that these new tests only test the previous timeout setting and do not meaningfully test the TCP keepalives; this is moreso just substantiating that we have, in fact, experienced a network interruption where we previously were not.

I am not marking this as draft because it improves our situation here, but maybe it could be even more improved.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/postgres-util/src/lib.rs
Comment (pH14) [inline src/postgres-util/src/lib.rs:443]: I've seen it recommended when using both keepalives and tcp user timeouts to set `tcp_user_timeout = keepalive_idle + keepalive_count*keepalive_interval` so an unanswered socket is dropped at the same time regardless of whether the sender had pending messages.

e.g. if we want a 60s for tcp user timeout, maybe we want to give the other end 5 attempts to respond via keep alive, so we set keepalive idle = 10, keepalive interval = 10, keepalive count = 5 or something like that.

File: src/postgres-util/src/lib.rs
Comment (pH14) [inline src/postgres-util/src/lib.rs:444]: I'm not sure how hard it is to wire in to this point, but it could be nice for these to be dynamically configurable so we could roll out changes without needing a new build

Comment (pH14) [reply src/postgres-util/src/lib.rs:444]: 60s is probably a safe starting point, but with the variability of the networking stacks sitting between us and all possible sources, I wonder if we'll want to fiddle with this value.

Comment (pH14) [reply src/postgres-util/src/lib.rs:444]: (this could also help the test case, where we could set the timeout very low and not have to wait 2 min)

**Comment by guswynn** [comment]
Will these take configs effect through an ssh tunnel?

File: src/postgres-util/src/lib.rs
Comment (sploiselle) [reply src/postgres-util/src/lib.rs]: ty for this config

**Comment by sploiselle** [comment]
I made the timeouts here configurable, but struggled to evoke any differences in behavior from any of the tests. Anyone have any pointers here?

**Review by benesch — COMMENTED** [review]
Re dynamic configuration: I _thiiiink_ @pH14 meant via Launch Darkly, not via end users? These knobs are so deep in the weeds that I don't think we'd ever want them to be user-settable.

**Comment by benesch** [comment]
I'm surprised the timeout toxic doesn't do the trick. Would have expected that to trigger exactly the same issue we're seeing in production, where the connection wedges but nothing in the source realizes.

**Comment by philip-stoev** [comment]
> Is there an existing pattern for only running mzcompose tests only on nightly? Glad to make a pattern but assumed there was something that I just couldn't find, so didn't want to make something needlessly duplicative.

I am sorry, I am afraid I do not understand the question. This mzcompose already runs in the normal CI -- what would you like to see run in Nightly?

**Review by philip-stoev — APPROVED** [review]
Thank you for fixing the original regression in the test!

It seems that you now have a test that is testing what you intended? Let me know if this is not the case and I will revisit this PR.

**Comment by benesch** [comment]
> It seems that you now have a test that is testing what you intended? Let me know if this is not the case and I will revisit this PR.

I think it is not testing what we observed in production, which was a hung TCP connection causing the PostgreSQL source to then hang for 25m, rather than noticing the hung connection and restarting.

**Comment by sploiselle** [comment]
Ok, these knobs turned out to not be necessary for this issue, but we already have them and tuning them might be useful in the future.

If the LaunchDarkly changes gets merged, I'll go through the final step of hooking the flags up. If this doesn't seem worthwhile, I'll pull the first few commits off to fix the bug in the `pg-cdc-resumption` tests.

cc @pH14

**Review by pH14 — APPROVED** [review]

**Comment by sploiselle** [comment]
I included a change that prevents us from waiting indefinitely on the PG stream; it's imperfect and could be made a little tighter to ensure we keep the right periodicity of heartbeats with PG but was simple and moves things in the right direction; will hold off on merging until I get an ok from @petrosagg or @benesch

https://github.com/MaterializeInc/materialize/pull/18644/commits/45dc6f854313fbd4591b10ab64c601497639f3f5

**Review by jkosh44 — APPROVED** [review]
Surfaces changes LGTM.

File: src/storage/src/source/postgres.rs
Comment (benesch) [inline src/storage/src/source/postgres.rs:1034]: I think in theory this should be `FEEDBACK_INTERVAL.saturating_sub(last_feedback.elapsed())`? But let's roll with what you have for now, since I'd like to backport it into v0.39.x.

**Review by benesch — APPROVED** [review]

File: src/sql/src/session/vars.rs
Comment (sploiselle) [reply src/sql/src/session/vars.rs:528]: Will do in a follow up

**Pull request merged**
